### ISSUE TITLE
Good training accuracy but low decode accuracy 

### ISSUE THREAD

**Issue by amadeusuzx** [espnet/espnet#3271]
Hi there!
I'm training a ASR task on my custom dataset using ESPnet 0.9.10. The accuracy on train set and val set was good, but when I decoded on test set, the accuracy was much more lower. Then I tried to decode on the val set and train set, and found the result quite similar ( acc about 20%). I posted the accuracy curve and decode results here. Could you help me find what's going wrong?
p.s. I extracted the feature using a custom encoder and used the wsj recipe for stage 2,3,4,5.

![image](https://user-images.githubusercontent.com/44519021/120748886-7a928900-c53e-11eb-91b7-bcd2ffbf92fc.png)

**Decode  result on test set.**
```
write a CER (or TER) result in exp/train_pytorch_train_no_preprocess/decode_test_decode_lm_word65000/result.txt
| SPKR    | # Snt    # Wrd   | Corr      Sub     Del     Ins     Err   S.Err  |
| Sum/Avg |  1218    32719   | 19.7      9.7    70.7     0.6    80.9   100.0  |
write a WER result in exp/train_pytorch_train_no_preprocess/decode_test_decode_lm_word65000/result.wrd.txt
|  SPKR    | # Snt     # Wrd  |  Corr     Sub      Del      Ins     Err    S.Err  |
|  Sum/Avg |  1218      6500  |   5.4    34.6     60.0      0.3    94.8    100.0  |
#example output
Scores: (#C #S #D #I) 1 3 8 0
REF:  DID THEY GET it DONE ON TIME AND FOR THE RIGHT MONEY
HYP:  *** **** *** it **** ** **** *** *** IS  NOT   SURE
Eval: D   D    D      D    D  D    D   D   S   S     S
```

**Comment by amadeusuzx** [comment]
My decode.yaml
```
batchsize: 0
beam-size: 10
penalty: 0.0
maxlenratio: 0.0
minlenratio: 0.0
ctc-weight: 0.3
lm-weight: 1.0
```
and train.yaml
```
# network architecture
# encoder related
transformer-input-layer: conv2d
elayers: 12
eunits: 2048
# decoder related
dlayers: 1
dunits: 1024
# attention related
adim: 256
aheads: 4
# transformer related
model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"

# hybrid CTC/attention
mtlalpha: 0.3

# label smoothing
lsm-weight: 0.1

# minibatch related
batch-size: 24 
maxlen-in: 512  # if input length  > maxlen_in, batchsize is automatically reduced
maxlen-out: 150 # if output length > maxlen_out, batchsize is automatically reduced

# optimization related
sortagrad: 0 # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
opt: noam
epochs: 120
dropout-rate: 0.1
accum-grad: 2
grad-clip: 5
patience: 0
transformer-lr: 20.0
transformer-warmup-steps: 25000
transformer-attn-dropout-rate: 0.0
transformer-length-normalized-loss: False
transformer-init: pytorch

```

**Comment by sw005320** [comment]
Hi @amadeusuzx,
What kind of tasks you're working on?
If it is the WSJ variant, the training/validation accuracy goes over 95%.
I'm concerning that your model is not well trained.
Then, this might be solved by changing the learning rate (`transformer-lr`), or you can revisit the feature that you extracted by yourself.

**Comment by amadeusuzx** [comment]
@sw005320 Thanks for your response!
I'm using e

**Comment by amadeusuzx** [comment]
@sw005320 Thanks for your response!
Actually I'm working on a visual-speech recognition task (lip reading) using the dataset [LRS2](https://www.robots.ox.ac.uk/~vgg/data/lip_reading/lrs2.html). The dataset consists of 25 fps videos and we encoded the frames to 512 dimensional vectors using an external encoder. There are about 48000 utterances in train set and 1000 utterances in both val and test set. I think the features extracted are correctly.
The weird thing is, the validation accuracy was about 75% during training, but when I decoded on the same val set, the character accuracy was only 20% (which is the same when I decoded in the test set).
Is that normal? Is the val accuracy computed in a different way from the decode stage?
Looking forward to your reply!

**Comment by sw005320** [comment]
I’m concerning that only the decoder part is trained. The accuracy is computed from the teacher forcing and it may happen.
Did you check that the recognition results are always same (it is not sure) or are there some variations?

I think @LiChenda was also working on this database.
@LiChenda, do you have some comments?

**Comment by sw005320** [comment]
Can you also remove the LM by setting `lm-weight` 0?
`lm-weight: 1` seems to be too high.

**Comment by amadeusuzx** [comment]
Oh certainly, the teacher forcing...

So how about deepening the encoder layer? Since The 512 dimensional features with high "frame stride"(i.e. low fps) are quite different from the fbank pitches, could you give me some advice about the encoder configuration?

And thanks for your suggestion of removing the LM. I'll try it soon!

**Comment by LiChenda** [comment]
Hello @amadeusuzx , I do you use `char` as the  modeling unit? In my experience, if you use `char` with the low fps (25Hz) visual features, the CTC loss (you can check the CTC loss curve to make sure) sometimes will not working (in some cases, the output char sequence will be longer than the input features).  In my experience, it's better to use BPE than char as the modeling unit.

> Oh certainly, the teacher forcing...
> 
> So how about deepening the encoder layer? Since The 512 dimensional features with high "frame stride"(i.e. low fps) are quite different from the fbank pitches, could you give me some advice about the encoder configuration?
> 
> And thanks for your suggestion of removing the LM. I'll try it soon!

**Comment by amadeusuzx** [comment]
@LiChenda 
Thanks for your reply! Yes I used characters as token and it seems like a bad idea.. I'll try bpe or phonemes and visemes. Btw, have you tried visual speech recognition on  LRS2 dataset using ESPnet? And if yes, can you share me your results or relevant publications?
Thanks so much for your help!

**Comment by LiChenda** [comment]
Yes, we recently have a paper accepted by interspeech 2021. Instead of the single talker AV-ASR, it's mainly focus on multi-talker audio-visual speech recognition. We haven't put it to arxiv due to some limitations, maybe my colleague @kqwyf can share you some numbers.
